[4.1] RavenDB-11296 Track Counters in session API

aviv86 opened the pull request:
CancellationTokens will be handled as a part of RavenDB-11326
http://issues.hibernatingrhinos.com/issue/RavenDB-11326

Still need to support includes with counters.

aviv86 committed:
RavenDB-11296 :
-Session Counters API : change from session.Advanced.Counters.Get(id, name) to session.CountersFor(id).Get(name)

aviv86 committed:
RavenDB-11296 : fix Counters tests after API change

aviv86 committed:
RavenDB-11296 : handle query on counters using session.CountersFor().Get()

aviv86 committed:
RavenDB-11296 : Add counters tracking to session

ayende commented on src/Raven.Client/Documents/Session/AsyncSessionDocumentCounters.cs:
```diff
@@ -0,0 +1,147 @@
+        {
+        }
+
+        public async Task<Dictionary<string, long>> GetAllAsync()
```
CancellationToken here? 
Do we have CT for all our methods?

ppekrol replied:
We don't. Wrote a Roslyn analyzer to check that.

ppekrol replied:
http://issues.hibernatingrhinos.com/issue/RavenDB-11326

ayende commented on src/Raven.Client/Documents/Session/AsyncSessionDocumentCounters.cs:
```diff
@@ -0,0 +1,147 @@
+            }
+
+            cache.GotAll = true;
+            Session.CountersByDocId[DocId] = cache;
```
This should be add _after_ the call, because an exception may be thrown here.

aviv86 replied:
fixed

ayende commented on src/Raven.Client/Documents/Session/AsyncSessionDocumentCounters.cs:
```diff
@@ -0,0 +1,147 @@
+                {
+                    Values = new Dictionary<string, long?>()
+                };
+                Session.CountersByDocId.Add(DocId, cache);
```
Only it add if after the call

aviv86 replied:
fixed

ayende commented on src/Raven.Client/Documents/Session/AsyncSessionDocumentCounters.cs:
```diff
@@ -0,0 +1,147 @@
+
+            if (Session.CountersByDocId.TryGetValue(DocId, out var cache))
+            {
+                if (cache.GotAll)
```
Merge the `GotAll` with the other call.
Then call to the server is didn't call `GotAll`

aviv86 replied:
fixed

ayende commented on src/Raven.Client/Documents/Session/DocumentSession.CountersFor.cs:
```diff
@@ -0,0 +1,25 @@
+            return new SessionDocumentCounters(this, documentId);
+        }
+
+        public SessionDocumentCounters CountersFor(object entity)
```
Is there a  reason we aren't resolving this to the `documentId` directly here?

ayende commented on src/Raven.Client/Documents/Session/ICountersSessionOperationsAsync.cs:
```diff
@@ -18,50 +17,18 @@ public interface ICountersSessionOperationsAsync : ICountersSessionOperationsBas
         /// <param name="counters">counters names</param>
         /// </summary>
-        Task<Dictionary<string, long>> GetAsync(object entity, params string[] counters);
+        Task<Dictionary<string, long?>> GetAsync(IEnumerable<string> counters);
```
I live how the interface shrank

ayende commented on src/Raven.Client/Documents/Session/InMemoryDocumentSessionOperations.cs:
```diff
@@ -120,6 +120,15 @@ public abstract partial class InMemoryDocumentSessionOperations : IDisposable
+            new Dictionary<string, CountersCache> (StringComparer.OrdinalIgnoreCase);
+
+        protected internal readonly HashSet<(string docId, string counter)> CountersToDelete = new HashSet<(string docId, string Name)>();
+        protected internal readonly HashSet<(string docId, string counter)> CountersToIncrement = new HashSet<(string docId, string Name)>();
```
What is this used for, and where is the actual amount to increment?

ayende replied:
Can we lazily allocate all of these? In many cases, they will not be used, and it will be extra garbage.

aviv86 replied:
removed both

ayende commented on src/Raven.Client/Documents/Session/InMemoryDocumentSessionOperations.cs:
```diff
@@ -749,6 +760,9 @@ internal SaveChangesData PrepareForSaveChanges()
 
             PrepareCompareExchangeEntities(result);
 
+            RemoveCountersFromCache(CountersToIncrement);
```
Why are you removing it? The result of the Inc operation will already tell you what the new value is, no?

aviv86 replied:
fixed, now we use the inc op result to update counters cache

ayende commented on src/Raven.Client/Documents/Session/InMemoryDocumentSessionOperations.cs:
```diff
@@ -820,6 +834,24 @@ private void PrepareCompareExchangeEntities(SaveChangesData result)
+                {
+                    // remove all
+                    CountersByDocId.Remove(docId);
+                    return;
```
I think you meant `continue;` here?

aviv86 replied:
i did :)

ayende commented on src/Raven.Client/Documents/Session/SessionCountersBase.cs:
```diff
@@ -13,16 +13,31 @@ namespace Raven.Client.Documents.Session
+        {
+            Session = session;
+
+            if (Session.DocumentsByEntity.TryGetValue(entity, out DocumentInfo document) == false)
```
Okay, so this is resolved once and we aren't holding on to it. Make sense.
However, I wonder if we can take advantage of this to skip the `GotAll` note. If we know that we have the metadata and are missing counters, we can check with the metadata in the session to know if we should get it.

aviv86 replied:
done

ppekrol commented on src/Raven.Client/Documents/Session/AsyncDocumentSession.cs:
```diff
@@ -101,9 +101,6 @@ protected override Task<string> GenerateIdAsync(object entity)
         public IRevisionsSessionOperationsAsync Revisions => _revisions ?? (_revisions = new DocumentSessionRevisionsAsync(this));
         private IRevisionsSessionOperationsAsync _revisions;
 
-        public ICountersSessionOperationsAsync Counters => _counters ?? (_counters = new DocumentSessionCountersAsync(this));
```
We do not want to have counters in .Advanced? It is a bit inconsistent with the rest of the session API.

ayende replied:
Yes, because it got really verbose and awkward when we started using it.
We also want to make this a 1st class item.
We'll probably also add `AttachmentsFor` to make this more discoverable

ppekrol commented on src/Raven.Client/Documents/Session/AsyncDocumentSession.CountersFor.cs:
```diff
@@ -0,0 +1,25 @@
+    /// </summary>
+    public partial class AsyncDocumentSession
+    {
+        public AsyncSessionDocumentCounters CountersFor(string documentId)
```
Just in case I would go with Interface here.

aviv86 replied:
we have one already https://github.com/ravendb/ravendb/pull/6549/files#diff-eefe503b561793713fc89e176ec41b71

ppekrol replied:
I mean as a return type of this method.

aviv86 replied:
done

aviv86 committed:
RavenDB-11296 PR fixes :
- update CountersByDocId only after server call
- remove CounterToDelete, CountersToIncrement hash sets
- update counters values in cache after increment operation
- SessionCounters now takes advantage of having an entity, by looking at the counters
  from metadata and registering the missing counters, in order to avoid GetAll when possible
- change return type of GetCountersFor from List to HashSet (we know that counter names are unique)
- update missing counters and deleted counters in cache after session Refresh and Load (not from cache)

aviv86 committed:
RavenDB-11296 : add more tests

aviv86 committed:
RavenDB-11296 : change return type of CountersFor() to interface

aviv86 committed:
RavenDB-11296 : avoid iterating over all the keys twice

ayende commented on src/Raven.Client/Documents/Commands/Batches/CountersBatchCommandData.cs:
```diff
@@ -30,6 +30,7 @@ public CountersBatchCommandData(string documentId, CounterOperation counterOpera
         public string Id { get; }
         public string Name { get; }
         public string ChangeVector { get; }
+        public bool HasIncrementOperation { get; set; }
```
Why do we need this property when we have `HasIncrement`?

aviv86 replied:
removed this property

ayende commented on src/Raven.Client/Documents/Linq/LinqPathProvider.cs:
```diff
@@ -468,7 +463,7 @@ public static string RemoveTransparentIdentifiersIfNeeded(string path)
         {
             return mce.Method.DeclaringType == typeof(RavenQuery) && mce.Method.Name == "Counter"
-                   || mce.Object?.Type == typeof(ICountersSessionOperations) && mce.Method.Name == "Get";
+                   || mce.Object?.Type == typeof(SessionDocumentCounters) && mce.Method.Name == "Get";
```
Shouldn't this be the interface here?

aviv86 replied:
fixed

ayende commented on src/Raven.Client/Documents/Session/AsyncSessionDocumentCounters.cs:
```diff
@@ -0,0 +1,153 @@
+
+            Dictionary<string, long> values;
+
+            if (cache.MissingCounters != null)
```
What is this for?
An in general, if we are already going to the server to get all, we might as well get them all and simplify the code.

aviv86 replied:
removed cache.MissingCounters,
if going to server will ask for all

ayende commented on src/Raven.Client/Documents/Session/InMemoryDocumentSessionOperations.cs:
```diff
@@ -120,6 +120,12 @@ public abstract partial class InMemoryDocumentSessionOperations : IDisposable
+        /// <summary>
+        /// hold the data required to manage Counters tracking for RavenDB's Unit of Work
+        /// </summary>
+        protected internal readonly Dictionary<string, CountersCache> CountersByDocId =
```
Use lazy creation for this, so we won't create a dictionary we won't need most times

aviv86 replied:
done

ayende commented on src/Raven.Client/Documents/Session/InMemoryDocumentSessionOperations.cs:
```diff
@@ -249,17 +255,17 @@ public void SetTransactionMode(TransactionMode mode)
+            if (documentInfo.Metadata.TryGet(Constants.Documents.Metadata.Counters, 
+                out BlittableJsonReaderArray counters) == false)
+                return null;
+            return new HashSet<string>(counters.Select(x=>x.ToString()), StringComparer.OrdinalIgnoreCase);
```
Is there a scenario where we _can_ have duplicate counter names?
And are counter names really case insensitive?

aviv86 replied:
reverted back to List

ayende commented on src/Raven.Client/Documents/Session/InMemoryDocumentSessionOperations.cs:
```diff
@@ -820,6 +830,22 @@ private void PrepareCompareExchangeEntities(SaveChangesData result)
 
+        private void PrepareForCountersIncrement(SaveChangesData result)
+        {
+            foreach (var (id, type, name) in result.DeferredCommandsDictionary.Keys)
```
Just iterate over the dictionary directly, not on the `Keys`, you can also then directly check the type of the value and not check the type.

aviv86 replied:
removed this method

ayende commented on src/Raven.Client/Documents/Session/InMemoryDocumentSessionOperations.cs:
```diff
@@ -1474,5 +1508,112 @@ public void OnBeforeQueryInvoke(BeforeQueryEventArgs beforeQueryEventArgs)
+            }
+
+            private bool _gotAll { get; set; }
+            private HashSet<string> _missingCounters;
```
What is the purpose of this?
Why not just put `Values[counter] = null` ?

aviv86 replied:
removed _missingCounters

ayende commented on src/Raven.Client/Documents/Session/Operations/BatchOperation.cs:
```diff
@@ -102,6 +112,24 @@ public void SetResult(BatchCommandResult result)
+        {
+            if (_session.CountersByDocId.TryGetValue(docId, out InMemoryDocumentSessionOperations.CountersCache cache) == false)
+            {
+                cache = new InMemoryDocumentSessionOperations.CountersCache();
```
You didn't _add_ it back to the `CountersByDocId` ?

aviv86 replied:
fixed

ayende commented on src/Raven.Client/Documents/Session/Operations/LoadOperation.cs:
```diff
@@ -115,6 +115,13 @@ public void SetResult(GetDocumentsResult result)
+
+                newDocumentInfo.Metadata.TryGet(Constants.Documents.Metadata.Counters, out BlittableJsonReaderArray bjra);
+                var metadataCounters = new HashSet<string>(bjra.Select(c => c.ToString()), StringComparer.OrdinalIgnoreCase);
+                cache.RegistarMissingCounters(metadataCounters);
```
What is the purpose of this code?

aviv86 replied:
removed

ayende commented on src/Raven.Client/Documents/Session/SessionCountersBase.cs:
```diff
@@ -13,16 +14,48 @@ namespace Raven.Client.Documents.Session
+            if (cache.GotAll == false && cache.MissingCounters == null)
+            {
+                var metadataCounters = Session.GetCountersFor(entity);
+                cache.RegistarMissingCounters(metadataCounters);
```
Why?

aviv86 replied:
removed cache.MissingCounters

ppekrol commented on src/Raven.Client/Documents/Operations/CountersOperationExecutor.cs:
```diff
@@ -81,14 +81,14 @@ public Task<CountersDetail> IncrementAsync(string docId, string name, long delta
         }
 
-        public Dictionary<string, long> Get(string docId, IEnumerable<string> counters)
+        public Dictionary<string, long?> Get(string docId, IEnumerable<string> counters)
```
Do we need CountersOperationExecutor at all? All now is a part of the session and we do not have equivalents for the rest. If the answer is 'yes' then we are missing there `SessionInfo sessionInfo = null` argument in each of those methods.

ppekrol replied:
Also all are OperationExecutors are based on IOperation or IServerOperation concept, but this one isnt.

ayende replied:
The one thing that we should be concerned here is the fact that:
* We need a lower level than the session
* Counters have different load balancing than other stuff, we likely will want to configure them separately.

ppekrol replied:
We can always use `store.Operations`.
If we want to have something different then I suggest renaming this to `DocumentStoreCounters` instead.

ayende replied:
@aviv86 Currently we _don't_ have a different behavior for load balancing on counters, right?

aviv86 replied:
right, we don't have a different behavior for load balancing on counters yet

aviv86 replied:
removed store.Counters
will recreate this in 4.2 with new policy
http://issues.hibernatingrhinos.com/issue/RavenDB-11348

aviv86 committed:
RavenDB-11296 : refactor counters cache

aviv86 committed:
RavenDB-11296 : fix PR comments

aviv86 committed:
RavenDB-11296 : fixes after rebase

aviv86 committed:
RavenDB-11296 : remove store.Counters

aviv86 committed:
RavenDB-11296 : use null propagation when addressing CountersByDocId from session

ppekrol commented on src/Raven.Client/Documents/Session/AsyncSessionDocumentCounters.cs:
```diff
@@ -0,0 +1,170 @@
+
+                Session.IncrementRequestCount();
+
+                var details = await Session.DocumentStore.Operations.ForDatabase(Session.DatabaseName)
```
I think that we have RequestExecutor and OperationExecutor in session already. Try using:
```
Session.Operations.SendAsync
```

Also remember to put `Session.SessionInfo` as a parameter to `SendAsync`.

ayende replied:
done

ppekrol commented on src/Raven.Client/Documents/Session/AsyncSessionDocumentCounters.cs:
```diff
@@ -0,0 +1,170 @@
+                Session.IncrementRequestCount();
+
+                var details = await Session.DocumentStore.Operations.ForDatabase(Session.DatabaseName)
+                    .SendAsync(new GetCountersOperation(DocId, new string[0]), token: token).ConfigureAwait(false);
```
I think it would be better from UX perspective to add a second constructor for this case.

Passing `null` or empty array should not be allowed. Check how it is done here: https://github.com/ravendb/ravendb/blob/43b7e45576d02df20df375eefc5b2da044bfb174/src/Raven.Client/Documents/Commands/GetDocumentsCommand.cs#L30-L51

ayende replied:
done

ayende commented on src/Raven.Client/Documents/Session/Operations/BatchOperation.cs:
```diff
@@ -102,6 +103,44 @@ public void SetResult(BatchCommandResult result)
+            for (int i = _sessionCommandsCount; i < result.Results.Length; i++)
+            {
+                if (!(result.Results[i] is BlittableJsonReaderObject batchResult))
+                    throw new ArgumentNullException();
```
You don't dictate what goes here, just ignore what you don't understand.
